locker/memory_locker.go

```diff
@@ -0,0 +1,153 @@
+package locker
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/featureform/fferr"
+	"github.com/google/uuid"
+)
+
+type memoryKey struct {
+	id      string
+	key     string
+	Channel *chan error
+}
+
+func (k *memoryKey) ID() string {
+	return k.id
+}
+
+func (k *memoryKey) Key() string {
+	return k.key
+}
+
+type MemoryLocker struct {
+	LockedItems sync.Map
+	Mutex       *sync.Mutex
+}
+
+func (m *MemoryLocker) Lock(key string) (Key, fferr.GRPCError) {
+	if key == "" {
+		return &memoryKey{}, fferr.NewInternalError(fmt.Errorf("cannot lock an empty key"))
+	}
+
+	m.Mutex.Lock()
+	defer m.Mutex.Unlock()
+
+	if m.isPrefixOfExistingKey(key) || m.hasPrefixLocked(key) {
+		return &memoryKey{}, fferr.NewKeyAlreadyLockedError(key, "", nil)
```

Comment:
passing in an empty string would result in a confusing log for `NewKeyAlreadyLockedError`
"key SomeKey is already locked by: "

Since this is prefix specific, maybe create a new error type or modify the existing so it reads:
"key SomeKey prefix is already locked"